cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -873,7 +874,8 @@ Result<std::vector<std::shared_ptr<Schema>>> ParquetDatasetFactory::InspectSchem
 
     size_t i = 0;
     for (const auto& e : paths_with_row_group_ids_) {
-      stripped[i++] = StripPrefixAndFilename(e.first, options_.partition_base_dir);
+      stripped[i++] =
+          StripPrefixAndFilename(e.first, options_.partition_base_dir).directory;
```

Comment:
Hmm, don't we still want the filename in the path in case the partitioning factory is a filename PartitioningFactory?

Reply:
This still seems off, but I can't figure out how to hit this case.

Reply:
I think we can do the same changes in the `Inspect()` method which currently accepts a path. Instead of passing a vector of strings, we can then pass a vector of `PartitionPathFormat` object, and then the `Inspect` methods of individual partitioning modes will use either the directory or the filename accordingly?

Reply:
That probably makes the most sense, but we might want to split out that refactoring separately, and also make sure that we can hit this path in a unit test in the first place (I was trying this morning but couldn't) as I don't want to expand the scope of this PR too much.

Reply:
The `Inspect()` methods now accept a `PartitionPathFormat` object as an argument with the latest commit. I have modified the tests and `SplitFilenameAndPrefix()` methods to return forms of PartitionPathFormat object as required. The `Inspect()` methods then extracts the directory or prefix accordingly whichever required.

Reply:
With the changes in the `Inspect()` method, I think the R build is failing, I am trying to investigate on fixing it, but not very sure about the R implementation.